py-polars/polars/datatypes.py

```diff
@@ -354,145 +364,171 @@ def __hash__(self) -> int:
-    tuple: List,
-    Decimal: Float64,
-    bytes: Binary,
-    object: Object,
-}
-
-_PY_STR_TO_DTYPE: dict[str, PolarsDataType] = {
-    str(tp.__name__): dtype for tp, dtype in _PY_TYPE_TO_DTYPE.items()
-}
-
-_DTYPE_TO_PY_TYPE: dict[PolarsDataType, type] = {
-    Float64: float,
-    Float32: float,
-    Int64: int,
-    Int32: int,
-    Int16: int,
-    Int8: int,
-    Utf8: str,
-    UInt8: int,
-    UInt16: int,
-    UInt32: int,
-    UInt64: int,
-    Boolean: bool,
-    Duration: timedelta,
-    Datetime: datetime,
-    Date: date,
-    Time: time,
-    Binary: bytes,
-    List: list,
-}
-
-# Map Numpy char codes to polars dtypes.
-#
-# Windows behaves differently from other platforms as C long is
-# only 32-bit on Windows, while it is 64-bit on other platforms.
-# See: https://numpy.org/doc/stable/reference/arrays.scalars.html
-_NUMPY_CHAR_CODE_TO_DTYPE = {
-    "b": Int8,
-    "h": Int16,
-    "i": Int32,
-    ("q" if sys.platform == "win32" else "l"): Int64,
-    "B": UInt8,
-    "H": UInt16,
-    "I": UInt32,
-    ("Q" if sys.platform == "win32" else "L"): UInt64,
-    "f": Float32,
-    "d": Float64,
-    "?": Boolean,
-}
-
-if _PYARROW_AVAILABLE:
-    _PY_TYPE_TO_ARROW_TYPE: dict[type, Callable[[], pa.lib.DataType]] = {
-        float: lambda: pyarrow_mod().float64(),
-        int: lambda: pyarrow_mod().int64(),
-        str: lambda: pyarrow_mod().large_utf8(),
-        bool: lambda: pyarrow_mod().bool_(),
-        date: lambda: pyarrow_mod().date32(),
-        time: lambda: pyarrow_mod().time64("us"),
-        datetime: lambda: pyarrow_mod().timestamp("us"),
-        timedelta: lambda: pyarrow_mod().duration("us"),
-    }
-
-    _DTYPE_TO_ARROW_TYPE: dict[PolarsDataType, Callable[[], pa.lib.DataType]] = {
-        Int8: lambda: pyarrow_mod().int8(),
-        Int16: lambda: pyarrow_mod().int16(),
-        Int32: lambda: pyarrow_mod().int32(),
-        Int64: lambda: pyarrow_mod().int64(),
-        UInt8: lambda: pyarrow_mod().uint8(),
-        UInt16: lambda: pyarrow_mod().uint16(),
-        UInt32: lambda: pyarrow_mod().uint32(),
-        UInt64: lambda: pyarrow_mod().uint64(),
-        Float32: lambda: pyarrow_mod().float32(),
-        Float64: lambda: pyarrow_mod().float64(),
-        Boolean: lambda: pyarrow_mod().bool_(),
-        Utf8: lambda: pyarrow_mod().large_utf8(),
-        Date: lambda: pyarrow_mod().date32(),
-        Datetime: lambda: pyarrow_mod().timestamp("us"),
-        Datetime("ms"): lambda: pyarrow_mod().timestamp("ms"),
-        Datetime("us"): lambda: pyarrow_mod().timestamp("us"),
-        Datetime("ns"): lambda: pyarrow_mod().timestamp("ns"),
-        Duration: lambda: pyarrow_mod().duration("us"),
-        Duration("ms"): lambda: pyarrow_mod().duration("ms"),
-        Duration("us"): lambda: pyarrow_mod().duration("us"),
-        Duration("ns"): lambda: pyarrow_mod().duration("ns"),
-        Time: lambda: pyarrow_mod().time64("us"),
-    }
+class _DataTypeMappings:
+    @property
+    @cache
```

Comment:
Maybe I am misunderstanding the usage of the LRU cache here, but this method always returns the same hard-coded dict, regardless of input arguments. Why would a cache speed this up?

Reply:
I think it doesn't have to reallocate every call and use the already allocated dict in the cache.

Reply:
The property  decorator makes it _look_ like an attribute, but it's still a function call; meaning without the cache you'd be recreating the dict on each access.

Reply:
Ah got it, it will be recreated, the cache avoids that. Thanks.